File: problems/CSES/Graph_Algorithms/course_schedule/course_schedule.py
import sys
sys.setrecursionlimit(200000)

input = sys.stdin.readline

def main():
    n, m = map(int, input().split())

    graph = [[] for _ in range(n)]
    for _ in range(m):
        a, b = map(int, input().split())
        graph[a - 1].append(b - 1)

    res = []
    visited = set()
    cycle = set()

    def dfs(course):
        if course in cycle:
            return False

        if course in visited:
            return True

        cycle.add(course)

        for nxt in graph[course]:
            if not dfs(nxt):
                return False

        cycle.remove(course)
        visited.add(course)

        res.append(course)

        return True

    for i in range(n):
        if i not in visited:
            if not dfs(i):
                print("IMPOSSIBLE")
                return

    res.reverse()

    print(*(x + 1 for x in res))

main()

File: problems/CSES/Graph_Algorithms/course_schedule/course_schedule.cpp
#include <bits/stdc++.h>
using namespace std;
using ll = long long;
#define pb push_back

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);

    int n, m;
    cin >> n >> m;

    vector<vector<int>> graph(n);

    for (int i = 0; i < m; i++) {
        int a, b;
        cin >> a >> b;

        graph[a - 1].pb(b - 1);
    }

    vector<int> res;

    unordered_set<int> visited;
    unordered_set<int> cycle;

    function<bool(int)> dfs = [&](int course) {
        if (cycle.count(course)) {
            return false;
        }

        if (visited.count(course)) {
            return true;
        }

        cycle.insert(course);

        for (int nxt : graph[course]) {
            if (!dfs(nxt)) {
                return false;
            }
        }

        cycle.erase(course);

        visited.insert(course);

        res.pb(course);

        return true;
    };

    for (int i = 0; i < n; i++) {
        if (!visited.count(i)) {
            if (!dfs(i)) {
                cout << "IMPOSSIBLE\n";

                return 0;
            }
        }
    }

    reverse(res.begin(), res.end());

    for (int x : res) {
        cout << x + 1 << ' ';
    }
    cout << '\n';

    return 0;
}
